### ISSUE TITLE
Separate data for scoped sub-routers?

### ISSUE THREAD

**Issue by seanpianka** [routerify/routerify#17]
Thanks for the update to add state to routers! It's an absolutely major feature that, in combination with the support for middleware and scopes, sets this crate apart from others. However, I'm running into unexpected behavior with setting data/state on scoped routers.

# Summary
`.data` can only be called on the top-most router. Any `.data` calls on scoped routers will not be set correctly, such that attempts to access via `req.data::<T>().unwrap()` will always return `None`.

# Expected Behavior

When I create a sub-router and add it as a scope within a parent router via `Router::builder().scope(...)`, the router passed into `scope(...)` should support having its own data.

In the below code example, I expect `GET /v1/users` to return `20` when the server below is being run. However, it panics when attempting to unwrap the data within the handler.

## Full Code Example

```rust
use hyper::{Body, Request, Response, Server};
// Import the routerify prelude traits.
use routerify::prelude::*;
use routerify::{Router, RouterService};
use std::io;
use std::net::SocketAddr;
use std::sync::{Arc, Mutex};

mod users {
    use super::*;

    struct State {
        count: Arc<Mutex<u8>>,
    }

    async fn list(req: Request<Body>) -> Result<Response<Body>, io::Error> {
        let count = req.data::<State>().unwrap().count.lock().unwrap();
        Ok(Response::new(Body::from(format!("Suppliers: {}", count))))
    }

    pub fn router() -> Router<Body, io::Error> {
        let state = State {
            count: Arc::new(Mutex::new(20)),
        };
        Router::builder()
            .data(state)
            .get("/", list)
            .build()
            .unwrap()
    }
}

mod offers {
    use super::*;

    struct State {
        count: Arc<Mutex<u8>>,
    }

    async fn list(req: Request<Body>) -> Result<Response<Body>, io::Error> {
        let count = req.data::<State>().unwrap().count.lock().unwrap();
        Ok(Response::new(Body::from(format!("Suppliers: {}", count))))
    }

    pub fn router() -> Router<Body, io::Error> {
        let state = State {
            count: Arc::new(Mutex::new(100)),
        };
        Router::builder()
            .data(state)
            .get("/", list)
            .build()
            .unwrap()
    }
}

#[tokio::main]
async fn main() {
    let scopes = Router::builder()
        .scope("/offers", offers::router())
        .scope("/users", users::router())
        .build()
        .unwrap();
    let router = Router::builder().scope("/v1", scopes).build().unwrap();
    // API Tree:
    //   /v1/users/ -> users::list   // returns 20
    //   /v1/offers/ -> offers::list // returns 100
    let service = RouterService::new(router).unwrap();
    let addr = SocketAddr::from(([127, 0, 0, 1], 3001));
    let server = Server::bind(&addr).serve(service);
    println!("App is running on: {}", addr);
    if let Err(err) = server.await {
        eprintln!("Server error: {}", err);
    }
}
```

# Output

The above code panics when making a request with this message:

```
thread 'tokio-runtime-worker' panicked at 'called `Option::unwrap()` on a `None` value', src/main.rs:40:21
```

Which is thrown from this line in the `offers::list` handler:

```
let count = req.data::<State>().unwrap().count.lock().unwrap();
```

# Solution

The solution here is to call `.data` on the top-most/parent router with the state required by all scopes. 

It would be nice if we could customize what `State` struct definition we use for each individual scope, since not all scopes require the same state.

## Full Code Example

```rust
use hyper::{Body, Request, Response, Server};
// Import the routerify prelude traits.
use routerify::prelude::*;
use routerify::{Router, RouterService};
use std::io;
use std::net::SocketAddr;
use std::sync::{Arc, Mutex};

mod users {
    use super::*;

    async fn list(req: Request<Body>) -> Result<Response<Body>, io::Error> {
        let count = req.data::<State>().unwrap().users_count.lock().unwrap();
        Ok(Response::new(Body::from(format!("Suppliers: {}", count))))
    }

    pub fn router() -> Router<Body, io::Error> {
        Router::builder().get("/", list).build().unwrap()
    }
}

mod offers {
    use super::*;

    async fn list(req: Request<Body>) -> Result<Response<Body>, io::Error> {
        let count = req.data::<State>().unwrap().offers_count.lock().unwrap();
        Ok(Response::new(Body::from(format!("Suppliers: {}", count))))
    }

    pub fn router() -> Router<Body, io::Error> {
        Router::builder().get("/", list).build().unwrap()
    }
}

struct State {
    offers_count: Arc<Mutex<u8>>,
    users_count: Arc<Mutex<u8>>,
}

#[tokio::main]
async fn main() {
    let state = State {
        offers_count: Arc::new(Mutex::new(100)),
        users_count: Arc::new(Mutex::new(20)),
    };
    let scopes = Router::builder()
        .scope("/offers", offers::router())
        .scope("/users", users::router())
        .build()
        .unwrap();
    let router = Router::builder()
        .data(state)
        .scope("/v1", scopes)
        .build()
        .unwrap();
    // API Tree:
    //   /v1/users/ -> users::list   // returns 20
    //   /v1/offers/ -> offers::list // returns 100
    let service = RouterService::new(router).unwrap();
    let addr = SocketAddr::from(([127, 0, 0, 1], 3001));
    let server = Server::bind(&addr).serve(service);
    println!("App is running on: {}", addr);
    if let Err(err) = server.await {
        eprintln!("Server error: {}", err);
    }
}
```

**Comment by seanpianka** [comment]
I see in `router/builder.rs:561`:

```rust
        for route in router.routes.iter_mut() {
            let new_route = Route::new_with_boxed_handler(
                format!("{}{}", path.as_str(), route.path.as_str()),
                route.methods.clone(),
                route.handler.take().expect("No handler found in one of the routes"),
            );
            builder = builder.and_then(move |mut inner| {
                inner.routes.push(new_route?);
                crate::Result::Ok(inner)
            });
        }
```

Perhaps, we could use an inner HashMap to map the scope's `path` to an `Option<T>`, where `T` is a type used for storing state. This way, the scope's state could then be later set with its `.data` method, but still be stored independently from other scopes.

**Comment by rousan** [comment]
it's my fault actually, i didn't notice it, scoped router will also have its own data. Thanks for pointing it out. Fix is on the way.

**Comment by seanpianka** [comment]
@rousan Thank you again! This isn't blocking my ability to use this crate anymore, as I have a functioning workaround, but with the addition of this feature/bugfix, I can achieve a code architecture very similar to the Go codebase that this project is being ported from. 

I hope this feature with stateful handlers makes this crate much more popular :) It's a much lighter weight and cleaner solution than something like actix_web or gotham, especially the middleware support.

**Comment by rousan** [comment]
Hey @seanpianka , I pushed the fix(scoped app state) in `scoped-data-support` branch, before making a new release, I want to make sure that everything is working.

So, can you please update the cargo dependency to github and test it out as follows:

```toml
[dependencies]
routerify = { git = "https://github.com/routerify/routerify", branch = "feature/scoped-data-support" }
```

and then use it as the examples you have mentioned above, now it will work.

```rust
use hyper::{Body, Request, Response, Server};
// Import the routerify prelude traits.
use routerify::prelude::*;
use routerify::{Router, RouterService};
use std::io;
use std::net::SocketAddr;
use std::sync::{Arc, Mutex};

mod users {
    use super::*;

    struct State {
        count: Arc<Mutex<u8>>,
    }

    async fn list(req: Request<Body>) -> Result<Response<Body>, io::Error> {
        let count = req.data::<State>().unwrap().count.lock().unwrap();
        Ok(Response::new(Body::from(format!("Suppliers: {}", count))))
    }

    pub fn router() -> Router<Body, io::Error> {
        let state = State {
            count: Arc::new(Mutex::new(20)),
        };
        Router::builder().data(state).get("/", list).build().unwrap()
    }
}

mod offers {
    use super::*;

    struct State {
        count: Arc<Mutex<u8>>,
    }

    async fn list(req: Request<Body>) -> Result<Response<Body>, io::Error> {
        let count = req.data::<State>().unwrap().count.lock().unwrap();
        Ok(Response::new(Body::from(format!("Suppliers: {}", count))))
    }

    pub fn router() -> Router<Body, io::Error> {
        let state = State {
            count: Arc::new(Mutex::new(100)),
        };
        Router::builder().data(state).get("/", list).build().unwrap()
    }
}

#[tokio::main]
async fn main() {
    let scopes = Router::builder()
        .scope("/offers", offers::router())
        .scope("/users", users::router())
        .build()
        .unwrap();
    let router = Router::builder().scope("/v1", scopes).build().unwrap();
    // API Tree:
    //   /v1/users/ -> users::list   // returns 20
    //   /v1/offers/ -> offers::list // returns 100
    let service = RouterService::new(router).unwrap();
    let addr = SocketAddr::from(([127, 0, 0, 1], 3001));
    let server = Server::bind(&addr).serve(service);
    println!("App is running on: {}", addr);
    if let Err(err) = server.await {
        eprintln!("Server error: {}", err);
    }
}
```

**Comment by rousan** [comment]
Even, you can access parent router's state from child router.

```rust
use hyper::{Body, Request, Response, Server};
// Import the routerify prelude traits.
use routerify::prelude::*;
use routerify::{Router, RouterService};
use std::io;
use std::net::SocketAddr;
use std::sync::{Arc, Mutex};

mod offers {
    use super::*;

    struct State {
        count: Arc<Mutex<u8>>,
    }

    async fn list(req: Request<Body>) -> Result<Response<Body>, io::Error> {
        let count = req.data::<State>().unwrap().count.lock().unwrap();

        // Access patent state
        println!("I can also access parent state: {:?}", req.data::<String>().unwrap());

        Ok(Response::new(Body::from(format!("Suppliers: {}", count))))
    }

    pub fn router() -> Router<Body, io::Error> {
        let state = State {
            count: Arc::new(Mutex::new(100)),
        };
        Router::builder().data(state).get("/", list).build().unwrap()
    }
}

#[tokio::main]
async fn main() {
    let scopes = Router::builder()
        .data("Parent State data".to_owned())
        .scope("/offers", offers::router())
        .build()
        .unwrap();
    let router = Router::builder().scope("/v1", scopes).build().unwrap();
    // API Tree:
    //   /v1/users/ -> users::list   // returns 20
    //   /v1/offers/ -> offers::list // returns 100
    let service = RouterService::new(router).unwrap();
    let addr = SocketAddr::from(([127, 0, 0, 1], 3001));
    let server = Server::bind(&addr).serve(service);
    println!("App is running on: {}", addr);
    if let Err(err) = server.await {
        eprintln!("Server error: {}", err);
    }
}

```

**Comment by seanpianka** [comment]
Thanks for the quick feature addition, it looks like everything with scoped data for sub-routers is working as intended! Below, I've prepared a full code sample for how I plan to use these scoped routers with their own state (along with the run-time initialization we discussed in #16).

However, I notice unexpected behavior:

When making a `GET /v1/users/` request, it responds as expected. However, `GET /v1/users` returns 404 Not Found. I don't think this is a bug because other crates offer middleware to "normalize" the route such that `/v1/users` and `/v1/users/` route to the same handler, but just something to note for the future.

# Full Code Sample
```rust
use hyper::{Body, Request, Response, Server};
use routerify::prelude::*;
use routerify::{Router, RouterService};
use std::{
    io,
    net::SocketAddr,
    sync::{Arc, Mutex},
};

mod users {
    use super::*;

    #[derive(Clone)]
    pub struct Data {
        pub count: Arc<Mutex<u8>>,
    }

    pub struct Handler {
        pub data: Data,
    }

    impl Handler {
        pub fn router(&self) -> Router<Body, io::Error> {
            Router::builder()
                .data(self.data.clone())
                .get("/", find_user)
                .build()
                .unwrap()
        }
    }

    async fn find_user(req: Request<Body>) -> Result<Response<Body>, io::Error> {
        let count = req.data::<Data>().unwrap().count.lock().unwrap();
        Ok(Response::new(Body::from(format!("Users: {}", count))))
    }
}

mod server {
    use super::*;
    use std::sync::{Arc, Mutex};

    pub struct Config {
        pub initial_user_count: u8,
    }

    pub struct Server {
        pub router: Router<Body, io::Error>,
    }

    impl Server {
        pub fn new(config: Config) -> Server {
            let users_handler = users::Handler {
                data: users::Data {
                    count: Arc::new(Mutex::new(config.initial_user_count)),
                },
            };
            let scopes = Router::builder()
                .scope("/users", users_handler.router())
                .build()
                .unwrap();
            let router = Router::builder().scope("/v1", scopes).build().unwrap();
            Server { router }
        }
    }
}

#[tokio::main]
async fn main() {
    let app = server::Server::new(server::Config {
        initial_user_count: 10,
    });
    let service = RouterService::new(app.router).unwrap();
    let addr = SocketAddr::from(([127, 0, 0, 1], 3001));
    let server = Server::bind(&addr).serve(service);

    println!("App is running on: {}", addr);
    if let Err(err) = server.await {
        eprintln!("Server error: {}", err);
    }
}
```

**Comment by seanpianka** [comment]
Also, in my above code example, if I change the route in `users::Handler::router` to:

```rust
.get("", find_user)
```

Then, `GET /v1/users`, it panics with a similar error as originally:

```
thread 'tokio-runtime-worker' panicked at 'called `Option::unwrap()` on a `None` value', src/main.rs:40:21
```

Perhaps the `path` argument being empty is an edge case? I was trying to write `#[test]`s for this library, but it seems that testing a Hyper server and handlers is also non-trivial for a Rust beginner.

**Comment by rousan** [comment]
Yeah, it's a bug and I was aware of this from the beginning. Thanks for reminding me. I will fix this with this release `v1.1.4`.

**Comment by rousan** [comment]
> thread 'tokio-runtime-worker' panicked at 'called `Option::unwrap()` on a `None` value', src/main.rs:40:21

While accessing the app data?

**Comment by seanpianka** [comment]
Yes, I'll post the full code example and panic stacktrace.

# Code Sample
```rust
use hyper::{Body, Request, Response, Server};
use routerify::prelude::*;
use routerify::{Router, RouterService};
use std::{
    io,
    net::SocketAddr,
    sync::{Arc, Mutex},
};

mod users {
    use super::*;

    #[derive(Clone)]
    pub struct Data {
        pub count: Arc<Mutex<u8>>,
    }

    pub struct Handler {
        pub data: Data,
    }

    impl Handler {
        pub fn router(&self) -> Router<Body, io::Error> {
            Router::builder()
                .data(self.data.clone())
                .get("", find_user)
                .build()
                .unwrap()
        }
    }

    async fn find_user(req: Request<Body>) -> Result<Response<Body>, io::Error> {
        let count = req.data::<Data>().unwrap().count.lock().unwrap();
        Ok(Response::new(Body::from(format!("Users: {}", count))))
    }
}

mod server {
    use super::*;
    use std::sync::{Arc, Mutex};

    pub struct Config {
        pub initial_user_count: u8,
    }

    pub struct Server {
        pub router: Router<Body, io::Error>,
    }

    impl Server {
        pub fn new(config: Config) -> Server {
            let users_handler = users::Handler {
                data: users::Data {
                    count: Arc::new(Mutex::new(config.initial_user_count)),
                },
            };
            let scopes = Router::builder()
                .scope("/users", users_handler.router())
                .build()
                .unwrap();
            let router = Router::builder().scope("/v1", scopes).build().unwrap();
            Server { router }
        }
    }
}

#[tokio::main]
async fn main() {
    let app = server::Server::new(server::Config {
        initial_user_count: 10,
    });
    let service = RouterService::new(app.router).unwrap();
    let addr = SocketAddr::from(([127, 0, 0, 1], 3001));
    let server = Server::bind(&addr).serve(service);

    println!("App is running on: {}", addr);
    if let Err(err) = server.await {
        eprintln!("Server error: {}", err);
    }
}
```

# Stacktrace

The following panic happens when making this request: `GET /v1/users`

```
thread 'tokio-runtime-worker' panicked at 'called `Option::unwrap()` on a `None` value', src/main.rs:33:21
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1069
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1427
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:218
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:511
  11: rust_begin_unwind
             at src/libstd/panicking.rs:419
  12: core::panicking::panic_fmt
             at src/libcore/panicking.rs:111
  13: core::panicking::panic
             at src/libcore/panicking.rs:54
  14: core::option::Option<T>::unwrap
             at /rustc/f509b26a7730d721ef87423a72b3fdf8724b4afa/src/libcore/macros/mod.rs:34
  15: routerifytest::users::find_user::{{closure}}
             at src/main.rs:33
```

**Comment by rousan** [comment]
It's throwing an error because the data is not available at `/users` path, but instead, it's available at `/users/*` path.

It's a bug I guess, This should be resolved internally. Will fix it with this release.

**Comment by rousan** [comment]
Do you think the both routes `/v1/users and /v1/users/ ` should work after resolve or it just be only `/v1/users` after it's resolved?

**Comment by seanpianka** [comment]
You're the best! Once I learn how to write `#[test]` for Hyper servers, I'll see about making a PR that can test these edge cases.

In my experience, it is typical behavior to have both `/v1/users` and /v1/users/` resolve to the same handler. However, I have also seen libraries/crates where this was left as an option to the developer. 

For example, actix_web provides a [NormalizePath](https://docs.rs/actix-web/3.0.0-alpha.3/actix_web/middleware/struct.NormalizePath.html) middleware. I found [this test](https://docs.rs/actix-web/3.0.0-alpha.3/src/actix_web/middleware/normalize.rs.html#114) which demonstrates the behavior:

```rust
    #[actix_rt::test]
    async fn test_wrap() {
        let mut app = init_service(
            App::new()
                .wrap(NormalizePath::default())
                .service(web::resource("/v1/something/").to(|| HttpResponse::Ok())),
        )
        .await;

        let req = TestRequest::with_uri("/v1//something////").to_request();
        let res = call_service(&mut app, req).await;
        assert!(res.status().is_success());

        let req2 = TestRequest::with_uri("//v1/something").to_request();
        let res2 = call_service(&mut app, req2).await;
        assert!(res2.status().is_success());

        let req3 = TestRequest::with_uri("//v1//////something").to_request();
        let res3 = call_service(&mut app, req3).await;
        assert!(res3.status().is_success());
    }
```

**Comment by rousan** [comment]
I think we should add an option to specify a path normalizer middleware to normalize crazy paths like ones above and the routerify will NOT provide any default normalizer as it might affect the performance unnecessarily, but the routerify will provide some basic path resolve e.g. `/users/` and `/users` will point to the same handler.

the normalizer could be assigned as below:

```rust
let router = Router::builder()
                // It will be responsible to normalize the request data e.g URI, method etc.
                // it will be called before any route processing.
                .normalizer(|req| async move { Ok(req) })
                .scope("/users", users_handler.router())
                .build()
                .unwrap();
```

what do you think?

**Comment by rousan** [comment]
or we can have a simple a `path_normalizer(|path: String| async move { Ok(path) })`.

**Comment by seanpianka** [comment]
Perhaps it could be a middleware, and the implementation is provided by the crate? For example:

```rust
// This middleware implementation could be provided by the crate.
async fn normalize_path_middleware(mut req: Request<Body>) -> Result<Response<Body>, io::Error> {
    // mutate/normalize the request here?
    Ok(res)
}

let router = Router::builder()
                .middleware(Middleware::pre(normalize_path_middleware))
                .scope("/users", users_handler.router())
                .build()
                .unwrap();
```

**Comment by seanpianka** [comment]
And I think benchmarking this path normalization would be helpful. If it doesn't affect performance by more than 1%, I would argue that making it default/opt-out behavior would make sense. However, it will be easier to just handle the typical case `/users/` and `/users`, and leave crazy path normalization for later.

**Comment by rousan** [comment]
Pushed the ` /users/ and /users` resolving in `develop` branch, please let me if everything is working, then I will make a new release.

Update the dep as follows:

```toml
[dependencies]
routerify = { git = "https://github.com/routerify/routerify", branch = "develop" }
```

**Comment by seanpianka** [comment]
Yep, looks good! My [example](https://github.com/routerify/routerify/issues/17#issuecomment-636350710) from before now returns the correct response from both `GET /v1/users` and `GET /v1/users/`. 

Thanks again for the fast response! This crate is shaping up nicely :)

**Comment by rousan** [comment]
Released [v1.1.4](https://github.com/routerify/routerify/releases/tag/v1.1.4)  🎉 .

Dep:

```toml
[dependencies]
routerify = "1.1"
```
